File: js/examples/06-reply-webhook.ts
/**
 * 06 · Handle a reply webhook
 *
 * A minimal receiver for Instantly webhook deliveries. Instantly provides NO
 * HMAC signature — the only delivery-auth is a custom header you set on the
 * webhook and verify here. The payload's `email_id` IS the `reply_to_uuid`,
 * so you can reply in-thread with replyToEmail.
 *
 * Register a webhook (once, needs a public URL) with registerWebhook() below,
 * then run this receiver:  PORT=3000 npm run 06
 */
import http from "node:http"
import { createInstantlyClient, createWebhook, replyToEmail, InstantlyApiError } from "@instantly-ai/sdk"

// --- One-time: point Instantly at your public receiver URL ---
export async function registerWebhook(apiKey: string, publicUrl: string, sharedSecret: string) {
  const client = createInstantlyClient({ apiKey })
  return createWebhook(client, {
    body: {
      target_hook_url: publicUrl,
      event_type: "reply_received",
      headers: { "x-shared-secret": sharedSecret }, // your only delivery-auth (no HMAC)
    },
  })
}

// --- The receiver ---
const SHARED_SECRET = process.env.WEBHOOK_SECRET ?? "change-me"
const port = Number(process.env.PORT ?? 3000)

const server = http.createServer((req, res) => {
  if (req.method !== "POST") {
    res.writeHead(405).end("method not allowed")
    return
  }
  let raw = ""
  req.on("data", (chunk) => (raw += chunk))
  req.on("end", async () => {
    if (req.headers["x-shared-secret"] !== SHARED_SECRET) {
      res.writeHead(401).end("bad secret")
      return
    }
    const event = JSON.parse(raw || "{}") as { event_type?: string; email_id?: string; reply_text?: string }
    // Match on the payload's event_type (may differ from the subscription enum).
    console.log(`event=${event.event_type} email_id=${event.email_id}`)

    // Example: to reply in-thread, event.email_id is the reply_to_uuid.
    // const apiKey = process.env.INSTANTLY_API_KEY!
    // await replyToEmail(createInstantlyClient({ apiKey }), {
    //   body: { reply_to_uuid: event.email_id!, eaccount: "you@acme.com", subject: "Re:", body: { html: "…" } },
    // })
    void replyToEmail // referenced so the import is meaningful in this template
    void InstantlyApiError

    res.writeHead(200).end("ok")
  })
})

server.listen(port, () => console.log(`Webhook receiver listening on :${port}`))

File: python/examples/06_reply_webhook.py
"""06 · Handle a reply webhook

A minimal receiver for Instantly webhook deliveries. Instantly provides NO HMAC
signature — the only delivery-auth is a custom header you set on the webhook and
verify here. The payload's `email_id` IS the `reply_to_uuid`, so you can reply
in-thread with reply_to_email.

Register once (needs a public URL) with register_webhook(), then run this receiver:
    PORT=3000 python 06_reply_webhook.py
"""
import json
import os
from http.server import BaseHTTPRequestHandler, HTTPServer

from instantly import create_instantly_client
from instantly.resources.webhook import create_webhook
from instantly.resources.email import reply_to_email  # noqa: F401  (used when you reply in-thread)

SHARED_SECRET = os.environ.get("WEBHOOK_SECRET", "change-me")


def register_webhook(api_key: str, public_url: str, shared_secret: str):
    """One-time: point Instantly at your public receiver URL."""
    client = create_instantly_client(api_key)
    return create_webhook(client, {
        "body": {
            "target_hook_url": public_url,
            "event_type": "reply_received",
            "headers": {"x-shared-secret": shared_secret},  # your only delivery-auth (no HMAC)
        },
    })


class Handler(BaseHTTPRequestHandler):
    def do_POST(self) -> None:  # noqa: N802
        if self.headers.get("x-shared-secret") != SHARED_SECRET:
            self.send_response(401)
            self.end_headers()
            self.wfile.write(b"bad secret")
            return
        length = int(self.headers.get("content-length", 0))
        event = json.loads(self.rfile.read(length) or b"{}")
        # Match on the payload's event_type (may differ from the subscription enum).
        print(f"event={event.get('event_type')} email_id={event.get('email_id')}")
        # To reply in-thread: event['email_id'] is the reply_to_uuid — pass it to reply_to_email.
        self.send_response(200)
        self.end_headers()
        self.wfile.write(b"ok")


def main() -> None:
    port = int(os.environ.get("PORT", "3000"))
    server = HTTPServer(("", port), Handler)
    print(f"Webhook receiver listening on :{port}")
    server.serve_forever()


if __name__ == "__main__":
    main()
